### PR TITLE
Final logits softcapping support for async GRPO Trainer

### LINKED ISSUE #5692: Add support for models using final_logits_softcapping to AsyncGRPOTrainer
### Feature request

Add support for models using final_logits_softcapping to AsyncGRPOTrainer

### Motivation

Some models (like gemma 2) use final_logits_softcapping and currently cannot be run with AsyncGRPOTrainer.

### Your contribution

I would be happy to collaborate on this feature request by creating a potential PR.

### PR DESCRIPTION
**Pull request by mlarnouhet**
# What does this PR do?

This PR patches patch_chunked_lm_head and _ChunkedLogProbFunction in trl/trainer/utils.py so that AsyncGRPOTrainer can support models using final_logits_softcapping (like gemma 2).

Fixes #5692

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> Touches the custom autograd path used to compute log-probs/entropy and gradients for chunked LM heads; small numerical or scaling mistakes could affect training stability/metrics on supported models.
> 
> **Overview**
> Adds support for models configured with `final_logit_softcapping` in the chunked LM head path used by async GRPO.
> 
> `_ChunkedLogProbFunction` now applies `logit_scale`, optional tanh-based softcapping, and temperature consistently in both forward and backward, and `patch_chunked_lm_head` passes the model’s `final_logit_softcapping` through instead of rejecting these models. Tests are updated to stop skipping softcapped models and to load bf16 models via `torch_dtype`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4f950e25bb3750eb9e0c6d65fa0ff7c115e0c1e3. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by AmineDiro** [comment]
HI @mlarnouhet , thanks for the contribution. That was a needed fix ! 

The changes look good, one thing missing is adding `trl-internal-testing/tiny-Gemma2ForCausalLM` family of models to  `/fsx/amine_dirhoussi/trl/tests/test_utils.py` so the tests can effectively run in :

https://github.com/huggingface/trl/blob/1240ecf4f00ca57ef0f158a3deb69d2299630ccd/tests/test_utils.py#L1230
and 
https://github.com/huggingface/trl/blob/1240ecf4f00ca57ef0f158a3deb69d2299630ccd/tests/test_utils.py#L1260

File: trl/trainer/utils.py [hunk 1]
```diff
@@ -1219,7 +1226,8 @@ def backward(ctx, grad_logprobs: torch.Tensor, grad_entropy: torch.Tensor):  # t
         temperature: float = ctx.temperature
         chunk_size: int = ctx.chunk_size
         logit_scale: float = ctx.logit_scale
-        inv_t = logit_scale / temperature
+        final_logit_softcapping: float = ctx.final_logit_softcapping
+        inv_t = 1 / temperature
```
Comment (AmineDiro) [inline trl/trainer/utils.py:1231]: same comment here for `logit_scale`

File: trl/trainer/utils.py [hunk 2]
```diff
@@ -1184,6 +1185,10 @@ def forward(
             torch.mm(last_hidden, w_chunk.t(), out=mm_buf[:, :C])
             logits_chunk = logits_buf[:, :C]
             logits_chunk.copy_(mm_buf[:, :C])
+
+            if final_logit_softcapping is not None:
+                logits_chunk.div_(final_logit_softcapping).tanh_().mul_(final_logit_softcapping)
+
```
[the remark this answers is not in the document]
Comment (AmineDiro) [reply trl/trainer/utils.py:1193]: order of scale matters bc we have tanh nonlinearity. Do we scale logits after the softcapping 🤔

[the remark this answers is not in the document]
Comment (mlarnouhet) [reply trl/trainer/utils.py:1193]: Both serve to prevent the output logits from being excessively large. The SFTTrainer's _chunk function applies logit_scale first so i did the same here. In the case where a model uses both, applying logit scale first could help prevent the logits from being truncated to +1 or -1 by the tanh, so i think it's better.

**Comment by AmineDiro** [comment]
Test looks good 👍🏼  let me rerun CI

**Review by AmineDiro — APPROVED** [review]
